### PR TITLE
W4 - D1- DinhTruong - AsyncTask + Thread + Handler

### PR DESCRIPTION
**Pull request by at-dinhtruong**
#23

### CODE REVIEW

File: app/src/at-dinhtruong/java/asiantech/internship/summer/asynctaskthreadhandler/AsyncTaskThreadHandlerActivity.java [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+import android.widget.Button;
+import android.widget.ImageView;
+
+import com.squareup.picasso.Picasso;
+
+import asiantech.internship.summer.R;
+
+public class AsyncTaskThreadHandlerActivity extends AppCompatActivity implements View.OnClickListener {
+    private ImageView mImgImageDownload;
+    String URL = "https://9mobi.vn/cf/images/2015/03/nkk/anh-dep-1.jpg";
```
Comment (at-dev-support-01) [inline app/src/at-dinhtruong/java/asiantech/internship/summer/asynctaskthreadhandler/AsyncTaskThreadHandlerActivity.java]: WARNING: [SuspiciousConstantFieldName] The field name indicates a constant but its modifiers do not
https://pmd.github.io/pmd-5.6.1/pmd-java/rules/java/naming.html#SuspiciousConstantFieldName

File: app/src/at-dinhtruong/res/layout/activity_async_task_thread_handler.xml [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+    android:layout_width="match_parent"
+    android:layout_height="match_parent"
+    android:orientation="vertical"
+    tools:context="asiantech.internship.summer.asynctaskthreadhandler.AsyncTaskThreadHandlerActivity">
+
+    <Button
+        android:id="@+id/btnDownLoadAsync"
+        android:layout_width="match_parent"
+        android:layout_height="wrap_content"
+        android:text="Download" />
```
Comment (at-dev-support-01) [inline app/src/at-dinhtruong/res/layout/activity_async_task_thread_handler.xml]: WARNING: [HardcodedText] Hardcoded string "Download", should use `@string` resource
 Hardcoding text attributes directly in layout files is bad for several reasons:

* When creating configuration variations (for example for landscape or portrait) you have to repeat the actual text (and keep it up to date when making changes)

* The application cannot be translated to other languages by just adding new translations for existing string resources.

There are quickfixes to automatically extract this hardcoded string into a resource lookup.

File: app/src/at-dinhtruong/java/asiantech/internship/summer/asynctaskthreadhandler/AsyncTaskThreadHandlerActivity.java [hunk 3]
```diff
@@ -0,0 +1,73 @@
+package asiantech.internship.summer.asynctaskthreadhandler;
+
+import android.annotation.SuppressLint;
```
Comment (at-dev-support-01) [inline app/src/at-dinhtruong/java/asiantech/internship/summer/asynctaskthreadhandler/AsyncTaskThreadHandlerActivity.java]: WARNING: [UnusedImports] Avoid unused imports such as 'android.annotation.SuppressLint'
https://pmd.github.io/pmd-5.6.1/pmd-java/rules/java/imports.html#UnusedImports

File: app/src/at-dinhtruong/java/asiantech/internship/summer/asynctaskthreadhandler/AsyncTaskThreadHandlerActivity.java [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+        }
+
+        protected Bitmap doInBackground(String... urls) {
+            String urldisplay = urls[0];
+            Bitmap bmp = null;
+            try {
+                InputStream inputStream = new URL(urldisplay).openStream();
+                bmp = BitmapFactory.decodeStream(inputStream);
+            } catch (Exception e) {
+                e.printStackTrace();
```
Comment (at-dev-support-01) [inline app/src/at-dinhtruong/java/asiantech/internship/summer/asynctaskthreadhandler/AsyncTaskThreadHandlerActivity.java]: WARNING: [AvoidPrintStackTrace] Avoid printStackTrace(); use a logger call instead.
https://pmd.github.io/pmd-5.6.1/pmd-java/rules/java/logging-java.html#AvoidPrintStackTrace

File: app/src/at-dinhtruong/java/asiantech/internship/summer/asynctaskthreadhandler/AsyncTaskThreadHandlerActivity.java [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+        mProgressDialog = new ProgressDialog(context);
+
+        mProgressDialog.setMessage("Downloading file..");
+        mProgressDialog.setIndeterminate(false);
+        mProgressDialog.setMax(100);
+        mProgressDialog.setProgressStyle(ProgressDialog.STYLE_HORIZONTAL);
+        mProgressDialog.setCanceledOnTouchOutside(false);
+    }
+*/
+    private class DownloadImageTask extends AsyncTask<String, Void, Bitmap> {
```
Comment (at-dev-support-01) [inline app/src/at-dinhtruong/java/asiantech/internship/summer/asynctaskthreadhandler/AsyncTaskThreadHandlerActivity.java]: WARNING: [StaticFieldLeak] This AsyncTask class should be static or leaks might occur (asiantech.internship.summer.asynctaskthreadhandler.AsyncTaskThreadHandlerActivity.DownloadImageTask)
 A static field will leak contexts.

Non-static inner classes have an implicit reference to their outer class. If that outer class is for example a `Fragment` or `Activity`, then this reference means that the long-running handler/loader/task will hold a reference to the activity which prevents it from getting garbage collected.

Similarly, direct field references to activities and fragments from these longer running instances can cause leaks.

ViewModel classes should never point to Views or non-application Contexts.

File: app/src/at-dinhtruong/java/asiantech/internship/summer/asynctaskthreadhandler/AsyncTaskThreadHandlerActivity.java [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+        DownloadTask();
+
+    }
+
+    @Override
+    public void onClick(View view) {
+        new DownloadImageTask().execute(URL_IMAGE);
+    }
+
+    public void DownloadTask() {
```
Comment (at-dev-support-01) [inline app/src/at-dinhtruong/java/asiantech/internship/summer/asynctaskthreadhandler/AsyncTaskThreadHandlerActivity.java]: ERROR: [MethodNamingConventions] Method names should not start with capital letters
https://pmd.github.io/pmd-5.6.1/pmd-java/rules/java/naming.html#MethodNamingConventions

File: app/src/at-dinhtruong/java/asiantech/internship/summer/asynctaskthreadhandler/AsyncTaskThreadHandlerActivity.java [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+                    if (isCancelled()) {
+                        input.close();
+                        return null;
+                    }
+                    total += count;
+                    if (fileLength > 0) {
+                        try {
+                            Thread.sleep(50);
+                        } catch (InterruptedException e) {
+                            e.printStackTrace();
```
Comment (at-dev-support-01) [inline app/src/at-dinhtruong/java/asiantech/internship/summer/asynctaskthreadhandler/AsyncTaskThreadHandlerActivity.java]: WARNING: [AvoidPrintStackTrace] Avoid printStackTrace(); use a logger call instead.
https://pmd.github.io/pmd-5.6.1/pmd-java/rules/java/logging-java.html#AvoidPrintStackTrace

File: app/src/at-dinhtruong/java/asiantech/internship/summer/asynctaskthreadhandler/AsyncTaskThreadHandlerActivity.java [hunk 8]
```diff
@@ -0,0 +1,107 @@
...
+    public void DownloadTask() {
+        mProgressDialog = new ProgressDialog(this);
+        mProgressDialog.setMessage("Downloading file..");
+        mProgressDialog.setIndeterminate(false);
+        mProgressDialog.setMax(100);
+        mProgressDialog.setProgressStyle(ProgressDialog.STYLE_HORIZONTAL);
+        mProgressDialog.setCanceledOnTouchOutside(false);
+    }
+
+    private class DownloadImageTask extends AsyncTask<String, Integer, Bitmap> {
```
Comment (at-dev-support-01) [inline app/src/at-dinhtruong/java/asiantech/internship/summer/asynctaskthreadhandler/AsyncTaskThreadHandlerActivity.java]: WARNING: [StaticFieldLeak] This AsyncTask class should be static or leaks might occur (asiantech.internship.summer.asynctaskthreadhandler.AsyncTaskThreadHandlerActivity.DownloadImageTask)
 A static field will leak contexts.

Non-static inner classes have an implicit reference to their outer class. If that outer class is for example a `Fragment` or `Activity`, then this reference means that the long-running handler/loader/task will hold a reference to the activity which prevents it from getting garbage collected.

Similarly, direct field references to activities and fragments from these longer running instances can cause leaks.

ViewModel classes should never point to Views or non-application Contexts.

File: app/src/at-dinhtruong/java/asiantech/internship/summer/asynctaskthreadhandler/AsyncTaskFragment.java [hunk 9]
```diff
@@ -0,0 +1,118 @@
...
+        btnDownloadAsync.setOnClickListener(this);
+    }
+
+
+    @Override
+    public void onClick(View view) {
+        new DownloadImageTask().execute(URL_IMAGE);
+    }
+
+    public void DownloadTask() {
```
Comment (at-dev-support-01) [inline app/src/at-dinhtruong/java/asiantech/internship/summer/asynctaskthreadhandler/AsyncTaskFragment.java]: ERROR: [MethodNamingConventions] Method names should not start with capital letters
https://pmd.github.io/pmd-5.6.1/pmd-java/rules/java/naming.html#MethodNamingConventions

File: app/src/at-dinhtruong/AndroidManifest.xml [hunk 10]
```diff
@@ -25,6 +29,7 @@
         <activity android:name=".activityandfragment.LoginActivity" />
         <activity android:name=".activityandfragment.InformationActivity" />
         <activity android:name=".recyclerview.RecyclerViewActivity" />
+        <activity android:name=".asynctaskthreadhandler.AsyncTaskThreadHandlerActivity" />
     </application>
 
-</manifest>
+</manifest>
```
Comment (at-quocnguyen) [inline app/src/at-dinhtruong/AndroidManifest.xml]: break line

File: app/src/at-dinhtruong/java/asiantech/internship/summer/asynctaskthreadhandler/AsyncTaskFragment.java [hunk 11]
```diff
@@ -0,0 +1,118 @@
...
+import java.io.InputStream;
+import java.net.HttpURLConnection;
+import java.net.URL;
+
+import asiantech.internship.summer.R;
+
+
+public class AsyncTaskFragment extends Fragment implements View.OnClickListener {
+    public static final String URL_IMAGE = "https://upload.wikimedia.org/wikipedia/commons/thumb/f/ff/Pizigani_1367_Chart_10MB.jpg/1280px-Pizigani_1367_Chart_10MB.jpg";
+    private ImageView mImgImageDownload;
```
Comment (at-quocnguyen) [inline app/src/at-dinhtruong/java/asiantech/internship/summer/asynctaskthreadhandler/AsyncTaskFragment.java]: Check lại name nè chỉ cần mImg...+ tên

File: app/src/at-dinhtruong/java/asiantech/internship/summer/asynctaskthreadhandler/HandlerFragment.java [hunk 12]
```diff
@@ -0,0 +1,60 @@
...
+import android.view.LayoutInflater;
+import android.view.View;
+import android.view.ViewGroup;
+import android.widget.Button;
+import android.widget.ImageView;
+
+import asiantech.internship.summer.R;
+
+public class HandlerFragment extends Fragment implements View.OnClickListener {
+    private ImageView mImgImageDownload;
```
Comment (at-quocnguyen) [inline app/src/at-dinhtruong/java/asiantech/internship/summer/asynctaskthreadhandler/HandlerFragment.java]: nt

File: app/src/at-dinhtruong/java/asiantech/internship/summer/asynctaskthreadhandler/HandlerFragment.java [hunk 13]
```diff
@@ -0,0 +1,60 @@
...
+    private Bitmap mBitmap;
+
+    @Override
+    public void onCreate(Bundle savedInstanceState) {
+        super.onCreate(savedInstanceState);
+
```
Comment (at-quocnguyen) [inline app/src/at-dinhtruong/java/asiantech/internship/summer/asynctaskthreadhandler/HandlerFragment.java]: ko dùng thì ko cần hoặc comment // no-op

File: app/src/at-dinhtruong/java/asiantech/internship/summer/asynctaskthreadhandler/ThreadFragment.java [hunk 14]
```diff
@@ -0,0 +1,67 @@
...
+import java.io.IOException;
+import java.io.InputStream;
+import java.net.HttpURLConnection;
+import java.net.URL;
+
+import asiantech.internship.summer.R;
+
+public class ThreadFragment extends Fragment implements View.OnClickListener {
+
+    private ImageView mImgImageDownload;
```
Comment (at-quocnguyen) [inline app/src/at-dinhtruong/java/asiantech/internship/summer/asynctaskthreadhandler/ThreadFragment.java]: nt

File: app/src/at-dinhtruong/java/asiantech/internship/summer/asynctaskthreadhandler/ThreadFragment.java [hunk 15]
```diff
@@ -0,0 +1,67 @@
...
+
+    @Override
+    public void onCreate(Bundle savedInstanceState) {
+        super.onCreate(savedInstanceState);
+
```
Comment (at-quocnguyen) [inline app/src/at-dinhtruong/java/asiantech/internship/summer/asynctaskthreadhandler/ThreadFragment.java]: check lại như trên

File: app/src/at-dinhtruong/res/layout/fragment_asynctask_handler_thread.xml [hunk 16]
```diff
@@ -0,0 +1,21 @@
...
+    tools:context="asiantech.internship.summer.asynctaskthreadhandler.AsyncTaskThreadHandlerActivity">
+
+    <Button
+        android:id="@+id/btnDownLoadAsync"
+        android:layout_width="match_parent"
+        android:layout_height="wrap_content"
+        android:text="@string/download" />
+
+    <ImageView
+        android:id="@+id/imgImageDownload"
```
Comment (at-quocnguyen) [inline app/src/at-dinhtruong/res/layout/fragment_asynctask_handler_thread.xml]: check lai name

**Comment by at-quocnguyen** [comment]
Resolve conflicts!

**Pull request merged**